static/js/typing.js

```diff
@@ -0,0 +1,107 @@
+// How long after someone stops editing in the compose box
+// do we send a 'stopped typing' notification
+var TYPING_STOPPED_WAIT_PERIOD = 5000; // 5s
+
+var users_currently_typing = [];
+
+function send_typing_notification_ajax(recipients, operation) {
+    channel.post({
+        url: '/json/typing',
+        data: {
+            to: recipients,
+            op: operation
+        },
+        success: function () {},
+        error: function (xhr, error_type) {
+            var response = channel.xhr_error_message("Error sending typing notification", xhr);
+        }
+    });
+}
+
+function check_and_send(operation) {
+    if (operation === 'start') {
+        if (compose.recipient() && compose.message_content()) {
+            send_typing_notification_ajax(compose.recipient(), operation);
+        }
+    } else if (operation === 'stop') {
+        if (compose.recipient()) {
+            send_typing_notification_ajax(compose.recipient(), operation);
+        }
+    }
+}
+
+function check_and_send_start() {
+    check_and_send('start');
+}
+
+function get_throttled_function() {
+    return _.throttle(check_and_send_start,
+        TYPING_STARTED_SEND_FREQUENCY,
+        {trailing: false});
+}
+
+function check_and_send_stop() {
+    check_and_send('stop');
+    exports.send_start_notification = get_throttled_function();
+    $(document).on('input', '#new_message_content', exports.send_start_notification);
+}
+
+exports.send_start_notification = get_throttled_function();
+exports.send_stop_notification = _.debounce(check_and_send_stop, TYPING_STOPPED_WAIT_PERIOD);
+
+$(document).on('compose_canceled.zulip', check_and_send_stop);
+$(document).on('compose_finished.zulip', check_and_send_stop);
+$(document).on('blur', '#new_message_content', check_and_send_stop);
+$(document).on('input', '#new_message_content', exports.send_start_notification);
+$(document).on('input', '#new_message_content', exports.send_stop_notification);
+
+function full_name(email) {
+    return people.get_by_email(email).full_name;
+}
+
+exports.hide_notification = function (event) {
+    // The typing notification is also sent to the user who is typing
+    // In this case the notification is not displayed
+    if (event.sender.user_id !== page_params.user_id) {
+        $('#typing_notifications').find('[data-email="' + event.sender.email + '"]').remove();
+        if (users_currently_typing.length === 1) {
+            $('#typing_notifications').hide();
+        }
+        var i = users_currently_typing.indexOf(event.sender.user_id);
+        users_currently_typing.splice(i);
+    }
+};
+
+var debounced_hide_notification = _.debounce(exports.hide_notification,
+  TYPING_STARTED_EXPIRY_PERIOD);
+
+exports.display_notification = function (event) {
+    // The typing notification is also sent to the user who is typing
+    // In this case the notification is not displayed
+    if (event.sender.user_id !== page_params.user_id) {
```

Comment:
I would write this check inverted using a `return` statement to avoid extra indentation in the following.